Submit example/llama with llama2/3 scripts and readme

lizamd opened the pull request.

lizamd committed:
Updated training script for llama2 and llama3. Added README.

gurpreet-dhami commented on examples/llama/train_llama3.sh:
@lizamd.
I had few changes to the llama2 script.
e.g.  Added a flag DISABLE_ROPE_TE
if [ "$DISABLE_ROPE_TE" -eq 1 ]; then
EXTRA_ARGS="$EXTRA_ARGS --disable-te-fused-rope"
fi

gurpreet-dhami replied:
@lizamd : Please add this flag. 
We added the support to enable/disable ROPE TE for Rocm through command line.

**Also I added the option to specify the DATA_DIR in case user wants to mention the specific path.
Because how will user copy the data in the docker at a specific location ?. It gets easier if user can specify the path** 
DATA_PATH=${DATA_PATH:-"$TMP_DIR/data/bookcorpus_text_sentence"}
https://github.com/ROCm/Megatron-LM/blob/d9a6c85a4e475b274cae6b0a0d58b22306c701a5/examples/llama2/train_llama2.sh#L58

**Same for checkpoint path**
CHECKPOINT_PATH=${CHECKPOINT_PATH:-"$TMP/ckpts"}
https://github.com/ROCm/Megatron-LM/blob/d9a6c85a4e475b274cae6b0a0d58b22306c701a5/examples/llama2/train_llama2.sh#L55C1-L55C49

**Can you take the llama2/train_llama2.sh and then add your changes on top of it  ?** 

@wenchenvincent , @lcskrishna : What do you say  ?

gurpreet-dhami commented on examples/llama/readme.md:
```diff
@@ -0,0 +1,146 @@
+   Download the Docker image required for training:  
+   `docker pull <image_name>`
+
+2. **Launch Docker Container**  
```
We added a Dockerfile for Rocm to create the docker image. So user can build the image as well.
https://github.com/ROCm/Megatron-LM/blob/rocm_dev/Dockerfile_rocm.ci

gurpreet-dhami commented on examples/llama/train_llama3.sh:
```diff
@@ -0,0 +1,335 @@
+fi
+
+
+echo 'import argparse
```
As you copied the mean_log_value.py file also from llama2 to llama folder. 
Lets Keep either this code or the file.

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -0,0 +1,335 @@
+TOTAL_ITERS="${TOTAL_ITERS:-10}"
+SEQ_PARALLEL="${SEQ_PARALLEL:-1}" 
+CONTI_PARAMS="${CONTI_PARAMS:-0}"
+OPTIMIZER="${OPTIMIZER:-sgd}"
```
This `OPTIMIZER` argument is not used anywhere. Either use it or remove it.

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -0,0 +1,338 @@
+TOTAL_ITERS="${TOTAL_ITERS:-5}"
+SEQ_PARALLEL="${SEQ_PARALLEL:-1}" 
+CONTI_PARAMS="${CONTI_PARAMS:-0}"
+OPTIMIZER="${OPTIMIZER:-sgd}"
```
This `OPTIMIZER` argument is not used anywhere. Either use it or remove it.

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -0,0 +1,338 @@
+        --lr-decay-style cosine \
+        --weight-decay 1.0e-1 \
+        --clip-grad 1.0 \
+        --optimizer sgd \
```
Change to Adam.

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -0,0 +1,335 @@
+        --lr-decay-style cosine \
+        --weight-decay 1.0e-1 \
+        --clip-grad 1.0 \
+        --optimizer sgd \
```
Change to Adam.

lcskrishna commented on examples/llama/train_llama2.sh:
```diff
@@ -0,0 +1,338 @@
+MASTER_ADDR="${MASTER_ADDR:-tw015}"
+MASTER_PORT="${MASTER_PORT:-6008}"
+#NNODES="${NNODES:-1}"
+NNODES="${NNODES:-2}"
```
Looks like the default setting is multi-node setting. Could you please use single node setting as default for this one?

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -0,0 +1,338 @@
+# MIT License
+#
+# Copyright (c) 2024 Advanced Micro Devices, Inc.
+#
```
Megatron-LM uses BSD 3-clause license. Please change accordingly.
You can refer to the license information here: https://github.com/ROCm/Megatron-LM/blob/rocm_dev/LICENSE#L5-L27

wenchenvincent replied:
Thanks for updating the license. We do still want to put the AMD copyright statement here.

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -0,0 +1,335 @@
+#
+# Copyright (c) 2024 Advanced Micro Devices, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
```
Megatron-LM uses BSD 3-clause license. Please change accordingly.
You can refer to the license information here: https://github.com/ROCm/Megatron-LM/blob/rocm_dev/LICENSE#L5-L27

wenchenvincent replied:
Thanks for updating the license. We do still want to put the AMD copyright statement here.

lcskrishna commented on examples/llama/train_llama2.sh:
```diff
@@ -0,0 +1,338 @@
+EXP_NAME="${EXP_NAME:-perf}"
+
+TEE_OUTPUT="${TEE_OUTPUT:-1}"
+NO_TORCH_COMPILE="${NO_TORCH_COMPILE:-1}"
```
I don't think ROCm/Megatron-LM supports TORCH_COMPILE.

lizamd committed:
modification based on review feedback

lizamd committed:
removed experiement/ folder

lizamd commented:
[AMD Official Use Only - AMD Internal Distribution Only]

Thanks for the feedback. Please note, the scripts are the ones we have been updating along with several customer projects. I am happy to update if yours have more features. I synced up with wen on merging your llama2 folder with this one so we don’t have 2 llama folders. All the tests have been done accordingly

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Gurpreet Singh Dhami ***@***.***>
Sent: Tuesday, November 19, 2024 9:17:31 AM
To: ROCm/Megatron-LM ***@***.***>
Cc: Li, Liz ***@***.***>; Mention ***@***.***>
Subject: Re: [ROCm/Megatron-LM] Submit example/llama with llama2/3 scripts and readme (PR #19)

Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


@gurpreet-dhami commented on this pull request.

________________________________

On examples/llama/train_llama3.sh<https://github.com/ROCm/Megatron-LM/pull/19#discussion_r1838637418>:

@lizamd<https://github.com/lizamd>.
I had few changes to the llama2 script.
e.g. Added a flag DISABLE_ROPE_TE
if [ "$DISABLE_ROPE_TE" -eq 1 ]; then
EXTRA_ARGS="$EXTRA_ARGS --disable-te-fused-rope"
fi

________________________________

In examples/llama/readme.md<https://github.com/ROCm/Megatron-LM/pull/19#discussion_r1838638930>:

> +
+---
+
+# Llama2/Llama3 Model Pretraining Instructions
+
+This guide provides the steps for setting up the environment and configuring the script to train Llama2 or Llama3 models.
+
+---
+
+## 1. Environment Setup
+
+1. **Download Docker Image**
+   Download the Docker image required for training:
+   `docker pull <image_name>`
+
+2. **Launch Docker Container**


We added a Dockerfile for Rocm to create the docker image. So user can build the image as well.
https://github.com/ROCm/Megatron-LM/blob/rocm_dev/Dockerfile_rocm.ci

________________________________

In examples/llama/train_llama3.sh<https://github.com/ROCm/Megatron-LM/pull/19#discussion_r1838643663>:

> +        $EXTRA_ARGS \
+        $TRAIN_ARGS \
+"
+
+if [ "$TEE_OUTPUT" -eq 0 ]; then
+    run_cmd="$run_cmd >& $TRAIN_LOG"
+else
+    run_cmd="$run_cmd |& tee $TRAIN_LOG"
+fi
+
+if [ "$NO_TRAINING" -eq 0 ]; then
+    eval $run_cmd
+fi
+
+
+echo 'import argparse


As you copied the mean_log_value.py file also from llama2 to llama folder.
Lets Keep either this code or the file.

________________________________

On examples/llama/train_llama3.sh<https://github.com/ROCm/Megatron-LM/pull/19#discussion_r1840847287>:

@lizamd<https://github.com/lizamd> : Please add this flag.
We added the support to enable/disable ROPE TE for Rocm through command line.

Also I added the option to specify the DATA_DIR in case user wants to mention the specific path.
Because how will user copy the data in the docker at a specific location ?. It gets easier if user can specify the path
DATA_PATH=${DATA_PATH:-"$TMP_DIR/data/bookcorpus_text_sentence"}
https://github.com/ROCm/Megatron-LM/blob/d9a6c85a4e475b274cae6b0a0d58b22306c701a5/examples/llama2/train_llama2.sh#L58

Same for checkpoint path
CHECKPOINT_PATH=${CHECKPOINT_PATH:-"$TMP/ckpts"}
https://github.com/ROCm/Megatron-LM/blob/d9a6c85a4e475b274cae6b0a0d58b22306c701a5/examples/llama2/train_llama2.sh#L55C1-L55C49

Can you take the llama2/train_llama2.sh and then add your changes on top of it ?

@wenchenvincent<https://github.com/wenchenvincent> , @lcskrishna<https://github.com/lcskrishna> : What do you say ?

—
Reply to this email directly, view it on GitHub<https://github.com/ROCm/Megatron-LM/pull/19#pullrequestreview-2430480914>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BGPJQJGBIA2B7CQBP6QMLND2BNXCXAVCNFSM6AAAAABRUVSF4CVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIMZQGQ4DAOJRGQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

gurpreet-dhami commented:
@lizamd : 
- As I see "This branch has conflicts that must be resolved". Can you please update this PR.
- I am fine with keeping your script. Please take necessary changes into your script.
- Please add this flag into your script - DISABLE_ROPE_TE. We added it to disable/enable TE Rope on Rocm through command line arguments.

- As you copied the mean_log_value.py file also from llama2 to llama folder. Lets Keep that code in llama2 bash script or the mean_log_value.py file.
- Also you can define checkpoint an data path like this "CHECKPOINT_PATH=${CHECKPOINT_PATH:-"$TMP/ckpts"}". so that user can give custom path.

lizamd committed:
modified scripts further

wenchenvincent commented on examples/llama/image-3.png:
Why is there nan in the snapshot?

lizamd committed:
updated readme

gurpreet-dhami commented on examples/llama/readme.md:
````diff
@@ -0,0 +1,131 @@
+### 2.1 Single Node Training
+To run the training on a single node, go to Megatron-LM folder, use the following command:
+```bash
+TEE_OUTPUT=1 MBS=5 BS=120 TP=8 TE_FP8=0 NO_TORCH_COMPILE=1 SEQ_LENGTH=4096 bash examples/llama/train_llama2.sh
````
please remove NO_TORCH_COMPILE option.

gurpreet-dhami commented on examples/llama/readme.md:
````diff
@@ -0,0 +1,131 @@
+
+- **On the Master Node:**
+  ```bash
+  TEE_OUTPUT=1 MBS=4 BS=64 TP=8 TE_FP8=0 NO_TORCH_COMPILE=1 SEQ_LENGTH=4096 bash examples/llama/train_llama2.sh
````
please remove NO_TORCH_COMPILE option.

gurpreet-dhami commented on examples/llama/readme.md:
````diff
@@ -0,0 +1,131 @@
+
+- **On the Slave Node(s):**
+  ```bash
+  TEE_OUTPUT=1 MBS=4 BS=64 TP=8 TE_FP8=0 NO_TORCH_COMPILE=1 SEQ_LENGTH=4096 bash examples/llama/train_llama2.sh
````
please remove NO_TORCH_COMPILE option.

gurpreet-dhami commented on examples/llama/train_llama2.sh:
```diff
@@ -0,0 +1,323 @@
+EXP_NAME="${EXP_NAME:-perf}"
+
+TEE_OUTPUT="${TEE_OUTPUT:-1}"
+#NO_TORCH_COMPILE="${NO_TORCH_COMPILE:-1}"
```
please remove the commented line  NO_TORCH_COMPILE.  we don't need this option anyways.

gurpreet-dhami committed:
resolve merge conflict

lizamd committed:
last update

lizamd committed:
modifed output format

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -0,0 +1,304 @@
+#!/bin/bash
+###############################################################################
+#################################################################################
```
Please add AMD copyright statement.

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -0,0 +1,305 @@
+#!/bin/bash
+###############################################################################
+#################################################################################
```
Please add AMD copyright statement.

lizamd committed:
add copyright
